### PR TITLE
idp_metadata_parser: consolidate signature verification key extraction

### PR DESCRIPTION
**Pull request by jgehrcke**
The current IdP metadata parser does not extract the `IDPSSODescriptor`'s `X509Certificate` value from https://www.testshib.org/metadata/testshib-providers.xml, because it applies an XPath expression that is too strict. This PR attempts to address this problem.

The commit msg:
```
- The XPath expression used by the metadata parser for extracting
  the key to be used for signature verification was too strict,
  as it required the `use=signing` attribute/value pair to be set,
  whereas the SAML spec declares the `use` attribute as optional.
  With this patch, parsing is retried with a relaxed XPath expression if
  the strict one did not yield a result.

- The parser now extracts a key from
  https://www.testshib.org/metadata/testshib-providers.xml
  The corresponding tests were adjusted to contain the `x509cert`
  key in the expected data structure.

- The parser now populates the `x509cert` value with a
  whitespace-free string, which is easier to test against
  and more reliable to program against.
```

@pitbulk would appreciate your review :-). Thank you! (Thanks for being so interactive, in general)

Edit: I should add that http://docs.oasis-open.org/security/saml/v2.0/saml-metadata-2.0-os.pdf page 15 says that `use` is optional.

### CODE REVIEW

**Comment by pitbulk** [comment]
> The parser now populates the `x509cert` value with a whitespace-free

And what happen with environments that expect those whitespaces?

**Comment by jgehrcke** [comment]
> And what happen with environments that expect those whitespaces?

Thanks for asking, and I understand that it is critical to properly asses behavioral change. This change is not the essential part of this PR, but I still believe that its a good change that should go in at this stage of the parser development.

I believe there are two important aspects:

#### 1) Parsers must ignore white space and other unexpected bytes

saml-metadata-2.0-os.pdf refers to https://www.w3.org/TR/xmldsig-core/#sec-X509Data as the specification of the what the `KeyDescriptor` element should contain. There we read:

> The X509Certificate element, which contains a base64-encoded [X509v3] certificate

Regarding white space treatment, the "base64-encoded" suggests (in practice, specifies) how white space should be treated. However, admittedly, that part of the spec puts us into one of those places inbetween Internet "standards" where we have to feel uncomfortable: The [X509v3] is misleading as it does not properly specify the word "base64-encoded". In the rest of the document, mentions of "base64" repeatedly refer to https://www.ietf.org/rfc/rfc2045.txt:

> The normative specification for base64 decoding transforms is [MIME]
 
which says in section 6.8:

> All line breaks or other characters not found in Table 1 must be ignored by decoding software.

and

> Any characters outside of the base64 alphabet are to be ignored in base64-encoded data.

In fact, the specification is not absolutely _precise_, but that's how, in practice, x509 certificate decoders work: they ignore the whitespace, because that's how "PEM-encoded x509 certificates" are to be decoded. The badness here is that the "PEM file format" is badly specified, also see http://stackoverflow.com/questions/5355046/where-is-the-pem-file-format-specified where an interesting point of view is:

> You are correct, it is not a real standard. When the SSLeay project (this later became the openssl project) needed a way to ASCII encode various public and private key files they were inspired by the PEM RFCs you cite and called the result a PEM file for lack of a better name. Now it refers to almost anything base64 encoded and wrapped with BEGIN and END lines. 

Fortunately, additional clarity comes from https://tools.ietf.org/html/rfc7468, which is a rather new RFC that accounts for the previously described specification imprecision (and now properly specifies the text encoding of certificates):

> Furthermore, parsers SHOULD ignore whitespace and other non-
   base64 characters

> The most lax parser implementations are not line-oriented at all and
   will accept any mixture of whitespace outside of the encapsulation
   boundaries

At the core, about all language environments should use one of the popular crypto libraries for actually decoding this data, and all popular crypto libraries I have seen ignore white space. In particular, all "PEM x509 cert parsers" I have seen (e.g. GnuTLS and OpenSSL) ignore that white space.

#### 2) I am under the impression that the IdP metadata parser is a rather recent and still fragile addition, so I believe it's okay to improve w/o fearing that consumers will want to punish us.

I hope to have clarified that actual interpretation of the data must not care about white space. Still, I understand your point about API consumers that 'expect' white space. Here, I am inclined to say: let's attempt to improve that parser component and make it easy to use as long as we still can :-). 

I suppose that we agree that the current parser state is very rudimentary and also highly error-prone, in view of the issue described here: in fact, the parser has so far been 'broken' for Shibboleth IdPs and also F5 (but probably many more IdPs) that do not make use of the `use` attribute.

So, I do believe the current nature of the parser justifies behavioral change. I believe it's not yet the time to be extremely conservative about behavioral change. But of course your mileage may vary in this regard

In conclusion, I think it's a good change. However, if you prever, we also could go back and expose the string literal that's found in the XML doc, unmodified.

I would like to note that we use that IdP XML metadata parser in DC/OS (thank you!), and that our intent will always be to send our patches upstream, to your project.

**Comment by jgehrcke** [comment]
@pitbulk is that segmentation fault error in the CI currently expected?
```
testSignMetadata (tests.src.OneLogin.saml2_tests.metadata_test.OneLogin_Saml2_Metadata_Test) ... /home/travis/build.sh: line 47:  6342 Segmentation fault      (core dumped)
```

**Comment by pitbulk** [comment]
Is an [issue on python-xmlsec](https://github.com/mehcode/python-xmlsec/issues/28#issuecomment-258713250), we trying to solve it now

**Comment by jgehrcke** [comment]
Thanks!
